File: JAVA/PartitionEqualSubsetSum.java
import java.util.*;

class Solution {
    public boolean canPartition(int[] nums) {
        int total = 0;
        for (int num : nums)
            total += num;
        if (total % 2 != 0)
            return false;
        return isPossible(nums, 0, total / 2, new HashMap<String, Boolean>());

    }

    private boolean isPossible(int[] nums, int currentIndex, int targetSum, HashMap<String, Boolean> memo) {
        if (targetSum == 0)
            return true;

        if (currentIndex >= nums.length)
            return false;

        String currentKey = Integer.toString(currentIndex) + "_" + Integer.toString(targetSum);

        if (memo.containsKey(currentKey))
            return memo.get(currentKey);

        boolean consider = false;
        int currentSum = nums[currentIndex];

        if (currentSum <= targetSum)
            consider = isPossible(nums, currentIndex + 1, targetSum - currentSum, memo);

        if (consider == true) {
            memo.put(currentKey, true);
            return true;
        }
        /*
         * if (currentSum <= targetSum){
         * consider = isPossible(nums, currentIndex + 1, targetSum - currentSum, memo);
         * 
         * if (consider == true) {
         * memo.put(currentKey, true);
         * return true;
         * }
         * }
         */
        boolean notconsider = isPossible(nums, currentIndex + 1, targetSum, memo);

        memo.put(currentKey, consider || notconsider);

        return memo.get(currentKey);
    }
}

File: C++/partitionEqualSubsetSum.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    bool canPartition(vector<int> &nums)
    {
        int sum = 0;
        for (int num : nums)
        {
            sum += num;
        }
        if (sum % 2 != 0)
        {
            return false;
        }
        map<string, bool> map;
        return isPossible(0, sum / 2, nums, map);
    }

private:
    bool isPossible(int currentIndex, int targetSum, vector<int> &nums, map<string, bool> &memo)
    {
        if (targetSum == 0)
            return true;

        if (currentIndex >= nums.size())
            return false;

        string currentKey = to_string(currentIndex) + "-" + to_string(targetSum);

        if (memo.find(currentKey) != memo.end())
            return memo[currentKey];

        int currentSum = nums[currentIndex];

        bool consider = false;

        if (currentSum <= targetSum)
        {
            consider = isPossible(currentIndex + 1, targetSum - currentSum, nums, memo);

            if (consider == true)
            {
                memo[currentKey] = true;
                return true;
            }
        }

        bool notConsider = isPossible(currentIndex + 1, targetSum, nums, memo);

        memo[currentKey] = consider || notConsider;

        return memo[currentKey];
    }
};

int main()
{
    Solution  sol;
    vector<int> nums = {1,5,11,5};
    cout  << sol.canPartition(nums);
    
}